src/SonarScanner.MSBuild.PreProcessor/ProcessedArgs.cs

```diff
@@ -225,32 +213,52 @@ private static bool CheckProjectKeyValidity(string key, ILogger logger)
         }
 
         // see spec in https://xtranet-sonarsource.atlassian.net/wiki/spaces/LANG/pages/3155001395/Scanner+Bootstrappers+implementation+guidelines
-        private SonarServer GetAndCheckSonarServer(ILogger logger, bool isHostSet, string sonarHostUrl, bool isSonarcloudSet, string sonarcloudUrl)
+        private ServerInfo GetAndCheckServerInfo(ILogger logger, bool isHostSet, string sonarHostUrl, bool isSonarcloudSet, string sonarcloudUrl)
         {
             const string defaultSonarCloud = "https://sonarcloud.io";
-            return new { isHostSet, isSonarcloudSet } switch
+            const string defaultSonarCloudApi = "https://api.sonarcloud.io";
+            var info = new { isHostSet, isSonarcloudSet } switch
             {
                 { isHostSet: true, isSonarcloudSet: true } when sonarHostUrl != sonarcloudUrl => Error(Resources.ERR_HostUrlDiffersFromSonarcloudUrl),
                 { isHostSet: true, isSonarcloudSet: true } when string.IsNullOrWhiteSpace(sonarcloudUrl) => Error(Resources.ERR_HostUrlAndSonarcloudUrlAreEmpty),
-                { isHostSet: true, isSonarcloudSet: true } => Warn(new SonarCloudServer(sonarcloudUrl), Resources.WARN_HostUrlAndSonarcloudUrlSet),
-                { isHostSet: false, isSonarcloudSet: false } => new SonarCloudServer(defaultSonarCloud),
+                { isHostSet: true, isSonarcloudSet: true } => Warn(new(sonarcloudUrl, defaultSonarCloudApi, true), Resources.WARN_HostUrlAndSonarcloudUrlSet),
+                { isHostSet: false, isSonarcloudSet: false } => new(defaultSonarCloud, defaultSonarCloudApi, true),
+                { isHostSet: false, isSonarcloudSet: true } => new(sonarcloudUrl, defaultSonarCloudApi, true),
                 { isHostSet: true, isSonarcloudSet: false } => sonarHostUrl == defaultSonarCloud
-                    ? new SonarCloudServer(defaultSonarCloud)
-                    : new SonarQubeServer(sonarHostUrl),
-                { isHostSet: false, isSonarcloudSet: true } => new SonarCloudServer(sonarcloudUrl),
+                    ? new(defaultSonarCloud, defaultSonarCloudApi, true)
+                    : new(sonarHostUrl, $"{sonarHostUrl.TrimEnd('/')}/api/v2", false),
             };
 
-            SonarServer Error(string message)
+            if (info is not null)
+            {
+                // Override by the user
+                var apiBaseUrl = AggregateProperties.TryGetProperty(SonarProperties.ApiBaseUrl, out var property)
+                    ? property.Value
+                    : info.ApiBaseUrl;
+
+                return new(info.ServerUrl, apiBaseUrl, info.IsSonarCloud);
```

Comment:
This is a bit ugly, since I am creating `ServerInfo` twice to indirectly mutate `ApiBaseUrl`.
Any ideas on how to do it differently?
I prefer this instead of adding a `set` on `ApiBaseUrl`, so that no one else can mutate it.

Reply:
We discussed offline and chose to keep this version.